File: lexer.py
import re
from wordlist import wordlist
from libparser import parcse_libs
import libbuilder

replacements = wordlist

variable_types = ['int', 'float', 'double', 'char', 'bool', 'string']
list_types = ['List', 'PPair']

libs, psclibs = parcse_libs()

def find_variables(tokens, classes):
    variables = {}
    for i in range(0, len(tokens)):
        if tokens[i] in variable_types:
            j = 1
            while tokens[i + j] == '*':
                j += 1
            variables[tokens[i+j]] = tokens[i]

        elif tokens[i] in list_types:
            j = 1
            if tokens[i+1].startswith('<') and tokens[i+1].endswith('>'):
                string = tokens[i+1].strip('<').strip('>') 
                
                j += 1
            while tokens[i + j] == '*':
                j += 1
            variables[tokens[i+j]] = tokens[i]

        elif tokens[i] in classes:
            j = 1
            while tokens[i + j] == '*':
                j += 1
            variables[tokens[i+j]] = tokens[i]

    return variables

def get_libs(tokens):
    includes = []
    for i in range(0, len(tokens)):
        if tokens[i] == 'friend':
            tokens[i+1] = tokens[i+1].strip('"')
            if tokens[i+1] in libs:
                includes.append(tokens[i+1])
                tokens[i+1] = f'"{tokens[i+1]}.h"'
                print(tokens[i+1], 2)

            elif tokens[i+1] in psclibs:
                libbuilder.build_lib(tokens[i+1])
                includes.append(tokens[i+1])
                tokens[i+1] = f'"{tokens[i+1]}.h"'
                print(tokens[i+1], 3)

            else: 
                tokens[i+1] = f'"{tokens[i+1]}.h"'
                print(tokens[i+1], 4)


    return includes


# def find_includes(tokens):
#     includes = []
#     for i in range(0, len(tokens)):
#         if tokens[i] == 'friend':
#             if tokens[i+1] in libs:
#                 tokens[i+1] = f'"{tokens[i+1]}.h"'
#             else: tokens[i+1] = f'"{tokens[i+1]}.h"'
#             includes.append(tokens[i+1])

    return includes

def find_classes(tokens:list):
    classes = []
    insert_indexes = []
    for i in range(0, len(tokens)-1):
        if tokens[i] == 'class':
            classes.append(tokens[i+1])
            j=0
            while tokens[i+j] != "{":
                j+=1
                if tokens[i+j] == '{':
                    insert_indexes.append(i+j+1) 

    updated_tokens = tokens
    for i in insert_indexes:
        tokens.insert(i, '\n')
        tokens.insert(i+1, 'public:')
            

    return classes, updated_tokens


def find_namespaces(tokens:list):
    namespaces = []
    insert_indexes = []
    for i in range(0, len(tokens)-1):
        if tokens[i] == 'namespace':
            namespaces.append(tokens[i+1])
            j=0
            while tokens[i+j] != "{":
                j+=1
                if tokens[i+j] == '{':
                    insert_indexes.append(i+j+1)


    return namespaces


def format_strings(tokens):
    for i in range(0, len(tokens)):
        if tokens[i-1] == 'f' and tokens[i].startswith('"') and tokens[i].endswith('"'):
            string_tokens = re.findall(r'\<.*?\>|\w+|{.*?}|[.,!?;:()\[\]=+\-*/<>]|', tokens[i].strip('"'))
            string = 'ponystring("")'
            for token in string_tokens:
                if token.startswith('{') and token.endswith('}'):
                    string += '+ ' + token.strip('{').strip('}')
                else:
                    string += '+ ' + f'ponystring("{token} ")'

            tokens[i] = string
            tokens[i-1] = ''
    return tokens      

def translate_tokens(tokens, variables, classes:list, namespaces:list, includes) -> list:
    for i in range(0, len(tokens)):
        try:
            if tokens[i] in variables:
                pass
            else:
                if tokens[i] == 'f' and tokens[i+1].startswith('"') and tokens[i+1].endswith('"'):
                    pass
                else:
                    if tokens[i+1] != '(' and tokens[i] not in variables and tokens[i] not in list_types and tokens[i] not in includes and tokens[i] != '\n' and tokens[i] not in ['+', '-', '*', '/', '=', '.', ',', ';', '(', ')', '[', ']', '{', '}', ':', '<', '>'] and not tokens[i].startswith('"') and not tokens[i].endswith('"') and not tokens[i].isdigit() and tokens[i] not in classes and not tokens[i].startswith('<') and not tokens[i].endswith('>') and tokens[i] not in namespaces:
                        if tokens[i] in replacements:
                            tokens[i] = replacements[tokens[i]]
                        else: 
                            raise Exception(f"Error: {tokens[i]} can`t be defined as command or object name")

        except IndexError: pass

    for i in range(0, len(tokens)):
        if tokens[i] in classes and tokens[i+1] == '.':
            tokens[i+1] = '::'
        if tokens[i] in namespaces and tokens[i+1] == '.':
            tokens[i+1] = '::'
    return tokens

def merge_float_tokens(tokens):
    merged_tokens = []
    i = 0
    while i < len(tokens):
        if re.match(r'\d+\.\d+', tokens[i]):
            # Начинаем объединение с текущего токена числа с плавающей точкой
            merged_token = tokens[i]
            i += 1
            # Продолжаем объединение, пока следующий токен также является числом с плавающей точкой
            while i < len(tokens) and re.match(r'\d+\.\d+', tokens[i]):
                merged_token += tokens[i]
                i += 1
            # Добавляем объединенный токен числа с плавающей точкой
            merged_tokens.append(merged_token)
        else:
            # Другие токены добавляем без изменений
            merged_tokens.append(tokens[i])
            i += 1
    return merged_tokens

def remove_comments(tokens):
    code = assemble_code(tokens)
    # Удаление однострочных комментариев
    code = re.sub(r'//.*', '', code)

    # Удаление многострочных комментариев
    code = re.sub(r'/\*(.*?)\*/', '', code, flags=re.DOTALL)

    return lexical_analyzer(code)

def lexical_analyzer(code):
    tokens = re.findall(r'\<.*?\>|\b(?:int|#include|using|namespace|void|return|string|public|private|protected|byte|bool|neigh|read|class)\b|\w+|".*?"|[.,!?;:(){}\[\]=+\-*/<>]|[\n]', code)
    tokens = ["#include" if item == "include" else item for item in tokens]
    return tokens

def assemble_code(tokens):
    assembled_code = ''
    last_token = ''
    for token in tokens:

        if token == '\n':
            assembled_code += '\n'
        elif token.startswith('"') and token.endswith('"') and last_token != '#include':
            assembled_code += ' ' + f'ponystring({token})'
        elif assembled_code.endswith('\n'):
            assembled_code += token
        elif token in ['+', '-', '*', '/', '=', '.', ',', ';', '(', ')', '[', ']', '{', '}', ':', '<', '>']:
            assembled_code += token
        elif token == 'using':
            assembled_code += token
        else:
            if token.isdigit() and last_token == '.':
                assembled_code += token
            else:
                assembled_code += ' ' + token
        last_token = token
    return assembled_code


def remove_comments(tokens):
    code = assemble_code(tokens)
    # Удаление однострочных комментариев
    code = re.sub(r'//.*', '', code)
    
    # Удаление многострочных комментариев
    code = re.sub(r'/\*(.*?)\*/', '', code, flags=re.DOTALL)
    
    return lexical_analyzer(code)

File: lexer.cpp
#include <regex>
#include <vector>
#include "utils.h"
#include "lexer.h"
#include "wordlist.h"

const std::vector<std::string> variable_types = {"int", "float", "double", "char", "bool", "string"};

std::vector<std::string> find_includes(std::vector<std::string> &tokens){
    std::vector<std::string> includes;
    for(int i = 0; i < tokens.size(); i++){
        if(tokens[i] == "friend"){
            includes.push_back(tokens[i+1]);
            tokens[i+1] = "\"" + tokens[i+1] + ".h" + "\"";
        }
    }
    return includes;
} 

std::vector<std::string> find_classes(std::vector<std::string> &tokens){
    std::vector<std::string> classes;
    std::vector<std::pair<int, std::string>> indexes_to_insert;
    

    for(int i = 0; i < tokens.size(); i++){
        if(tokens[i] == "class"){
          classes.push_back(tokens[i+1]);
          int arguments_start = i+2;
          int arguments_end = i+2;
          int j =0;
          while(i+j < tokens.size()){
            j++;
            if(tokens[i+j] == "{"){
                indexes_to_insert.push_back(std::make_pair(i + j, "public: \n"));
                arguments_end = i+j-1;
            }
            std::string arguments = "";
            for(int k = arguments_start; k <= arguments_end; k++){
                arguments += tokens[k] + " ";
            }
            indexes_to_insert.push_back(std::make_pair(i + j+1, arguments));
            arguments.clear();

          }
        }
    }


    for(int i = 0; i < indexes_to_insert.size(); i++){
        tokens.insert(tokens.begin() + indexes_to_insert[i].first, indexes_to_insert[i].second);
    }

    indexes_to_insert.clear();

    return classes;
}

std::vector<std::string> find_namespaces(std::vector<std::string> &tokens){
    std::vector<std::string> namespaces;
    for(int i = 0; i < tokens.size(); i++){
        if(tokens[i] == "namespace"){
          namespaces.push_back(tokens[i+1]);
        }
    }
    return namespaces;
}

std::vector<std::string> find_functions(std::vector<std::string> &tokens){
    std::vector<std::string> functions;
    for(int i = 0; i < tokens.size(); i++){
        if(tokens[i] == "("){
          functions.push_back(tokens[i-1]);
        }
    }
    return functions;
}

std::vector<std::string> find_variables(std::vector<std::string> &tokens, std::vector<std::string> classes){
    std::vector<std::string> variables;
    for(int i = 0; i < tokens.size(); i++){
        if(std::find(variable_types.begin(), variable_types.end(), tokens[i]) != variable_types.end()){
            int j = 1;
            while(tokens[i+j] == "*"){
                j++;
            }
            variables.push_back(tokens[i+j]);
        } 
        else if(std::find(classes.begin(), classes.end(), tokens[i]) != classes.end()){
            int j = 1;

            if(tokens[i+1] == "<"){
                while(tokens[i+j] != ">"){
                    j++;
                }
            }

            while(tokens[i+j] == "*"){
                j++;
            }
            variables.push_back(tokens[i+j]);
        }

    }
    return variables;
}

std::vector<std::string> format_strings(std::vector<std::string> &tokens){
    std::vector<std::string> updated_tokens = tokens;
    std::vector<int> indexes_to_remove;
    std::vector<int> indexes_to_insert;

    for(int i = 0; i < tokens.size(); i++){

        if(is_string(tokens[i]) && tokens[i-1] == "f"){
            std::string str = strip(tokens[i], '\"');
            std::string formatted_str = "\"\"";

            int last_index = 0;
            for(int j = 0; j < str.size(); j++){
                if(str[j] == '{'){
                    formatted_str  += " + \"" + str.substr(last_index, j - last_index) +"\"";
                    int k = j;
                    while(k < str.size() && str[k] != '}'){
                        k++;
                    }
                    std::string code_string = str.substr(j + 1, k - j - 1);
                    std::vector<std::string> code_tokens = lexical_analyzer(code_string);
                    std::string code_str = assemble_tokens(format_strings(code_tokens));
                    formatted_str += " + ponystring(" + code_str + ")";
                    last_index = k + 1;
                }
            }

            formatted_str = " ponystring(" + formatted_str + ")";



            updated_tokens[i] = formatted_str;
            indexes_to_remove.push_back(i-1);
        }
    }

    for(int i = 0; i < indexes_to_remove.size(); i++){
        updated_tokens.erase(updated_tokens.begin() + indexes_to_remove[i]);
    }

    tokens = updated_tokens;
    return tokens;
}

std::vector<std::string> translate_tokens(std::vector<std::string> &tokens, std::vector<std::string> variables, std::vector<std::string> classes, std::vector<std::string> namespaces, std::vector<std::string> includes){
    std::vector<std::string> symbols = {"+", "-", "*", "/", "=", ".", ",", ";", "(", ")", "[", "]", "{", "}", ":", "<", ">"};
    for(int i = 0; i < tokens.size(); i++){
        if (!is_vector_contain_string(variables, tokens[i]) && tokens[i+1] != "(" && !is_vector_contain_string(classes, tokens[i]) 
        && !is_vector_contain_string(namespaces, tokens[i]) && !is_vector_contain_string(includes, tokens[i]) 
        && !is_vector_contain_string(symbols, tokens[i]) && tokens[i] != "\n" && !isDigit(tokens[i]) && !is_string(tokens[i]) && !is_type(tokens[i])) 
        {
            if(wordlist.find(tokens[i]) != wordlist.end()){
                tokens[i] = wordlist.at(tokens[i]);
            }
        }
    }
    return tokens;
}

std::vector<std::string> lexical_analyzer(std::string str)
{
std::regex reg("<.*?>|\\b(?:int|#include|using|namespace|void|return|string|public|private|protected|byte|bool|neigh|read|class)\\b|\\w+|\".*?\"|[.,!?;:(){}\\[\\]=+\\-*/<>]|[\n]");

std::sregex_iterator begin(str.begin(), str.end(), reg);
std::sregex_iterator end;
std::vector<std::string> tokens;
for (std::sregex_iterator it = begin; it != end; ++it) {
    std::smatch match = *it;
    tokens.push_back(match.str());
}
return tokens;
}


std::string assemble_tokens(std::vector<std::string> tokens){
    std::string assembled_code = "";
    std::string last_token = "";
    std::vector<std::string> symbols = {"+", "-", "*", "/", "=", ".", ",", ";", "(", ")", "[", "]", "{", "}", ":", "<", ">"};

    for(std::string token : tokens){

        if(token == "\n")
            assembled_code += "\n";
        else if (is_string(token) && last_token != "#include")
            assembled_code += "ponystring(" + token + ")";
        else if (assembled_code[-1] == '\n')
            assembled_code += token;
        else if(std::find(symbols.begin(), symbols.end(), token) != symbols.end())
            assembled_code += token;
        else if(isDigit(token) && last_token == ".")
            assembled_code += " " + token;
        else 
            assembled_code += " " + token;

        last_token = token;

    }
    return assembled_code;
}

std::vector<std::string> remove_comments(const std::vector<std::string>& tokens) {
    std::string code = assemble_tokens(tokens);
    code = std::regex_replace(code, std::regex("//.*"), "");
    code = std::regex_replace(code, std::regex("/\\*(.*?)\\*/"), "", std::regex_constants::format_first_only);
    return lexical_analyzer(code);
}


